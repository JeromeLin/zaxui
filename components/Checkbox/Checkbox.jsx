import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Cell from '../Cell';
import Button from '../Button';
import Icon from '../Icon';

class Checkbox extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: props.checked || props.defaultChecked || false,
    };
    this.onValueChange = this.onValueChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: nextProps.checked || nextProps.defaultChecked || false,
      });
    }
  }

  onValueChange() {
    const { disabled, onChange } = this.props;

    if (disabled) {
      return;
    }

    const checked = !this.state.checked;
    this.setState({ checked });
    typeof onChange === 'function' && onChange(checked);
  }

  render() {
    const { prefixCls, className, theme, type, value, block, disabled, id, children } = this.props;
    const { checked } = this.state;

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
      checked,
      disabled,
    });

    if (type === 'cell') {
      return (
        <Cell disabled={disabled} description={checked ? <Icon type="right" theme={disabled ? null : theme} /> : null} onClick={() => {}}>
          <input type="checkbox" className={`${prefixCls}-input`} value={value} disabled={disabled} checked={checked} onChange={this.onValueChange} />
          {children}
        </Cell>
      );
    }

    if (type === 'button') {
      return (
        <Button className={cls} theme={theme} size="xs" block={block} bordered={!checked} disabled={disabled}>
          <input type="checkbox" className={`${prefixCls}-input`} disabled={disabled} checked={checked} onChange={this.onValueChange} />
          {children}
        </Button>
      );
    }

    return (
      <div className={cls}>
        <div className={`${prefixCls}-wrapper`}>
          <span className={`${prefixCls}-inner`} />
          {
            children
              ? <span className={`${prefixCls}-text`}>{children}</span>
              : null
          }
          <input id={id} type="checkbox" className={`${prefixCls}-input`} disabled={disabled} checked={checked} onChange={this.onValueChange} />
        </div>
      </div>
    );
  }
}

Checkbox.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  type: PropTypes.oneOf(['button', 'cell']),
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
};

Checkbox.defaultProps = {
  prefixCls: 'za-checkbox',
  className: null,
  theme: 'primary',
  type: null,
  defaultChecked: false,
  disabled: false,
  onChange() {},
};

export default Checkbox;
