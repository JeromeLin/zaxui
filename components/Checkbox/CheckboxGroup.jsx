import React, { PureComponent, cloneElement } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function getCheckedValue(children) {
  const checkedValue = [];
  React.Children.forEach(children, (checkbox) => {
    if (checkbox.props && checkbox.props.checked) {
      checkedValue.push(checkbox.props.value);
    }
  });
  return checkedValue;
}

class CheckboxGroup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue || getCheckedValue(props.children),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps || getCheckedValue(nextProps.children)) {
      this.setState({
        value: nextProps.value || nextProps.defaultValue || getCheckedValue(nextProps.children),
      });
    }
  }

  onCheckboxChange(value) {
    const values = this.state.value;
    const index = values.indexOf(value);

    if (index < 0) {
      values.push(value);
    } else {
      values.splice(index, 1);
    }

    this.setState({
      value: values,
    });
    this.props.onChange(values);
  }

  render() {
    const { prefixCls, className, theme, shape, type, block, disabled, compact, children } = this.props;

    const items = React.Children.map(children, (element, index) => {
      return cloneElement(element, {
        key: index,
        type,
        theme,
        block: block || element.props.block,
        disabled: disabled || element.props.disabled,
        onChange: () => this.onCheckboxChange(element.props.value),
        checked: (this.state.value.indexOf(element.props.value) > -1),
      });
    });

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
      [`shape-${shape}`]: !!shape,
      'is-compact': compact,
      block,
      disabled,
    });

    return (
      <div className={cls}>
        {items}
      </div>
    );
  }
}

CheckboxGroup.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  shape: PropTypes.oneOf(['radius', 'round']),
  type: PropTypes.oneOf(['button', 'cell']),
  value: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
  defaultValue: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
  block: PropTypes.bool,
  disabled: PropTypes.bool,
  compact: PropTypes.bool,
  onChange: PropTypes.func,
};

CheckboxGroup.defaultProps = {
  prefixCls: 'za-checkbox-group',
  className: null,
  theme: 'primary',
  shape: null,
  type: null,
  value: null,
  defaultValue: null,
  block: false,
  disabled: false,
  compact: false,
  onChange() {},
};

export default CheckboxGroup;
