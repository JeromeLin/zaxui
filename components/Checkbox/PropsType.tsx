interface BaseCheckboxProps {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  size?: 'lg';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (checked: boolean) => void;
  id?: string;
  checked?: boolean;
  defaultChecked?: boolean;
  value?: number | string;
}

export interface CheckboxProps extends BaseCheckboxProps {
  prefixCls?: string;
  className?: string;
}


interface BaseCheckboxGroupProps {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  size?: 'lg';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (checked: boolean) => void;
  value?: Array<number | string>;
  defaultValue?: Array<number | string>;
  compact?: boolean;
}

export interface CheckboxGroupProps extends BaseCheckboxGroupProps {
  prefixCls?: string;
  className?: string;
}