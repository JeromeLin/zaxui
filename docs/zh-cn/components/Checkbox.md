# 复选框 Checkbox

[demo页面](https://zhongantecheng.github.io/zarm/#/checkbox)

### 引入

```js
import { Checkbox } from 'zarm';
```

### 代码演示

#### 基本用法

单独使用

```jsx
<Checkbox onChange={checked => console.log(`checkbox to ${checked}`)}>同意条款</Checkbox>
```

组合使用

```jsx
<Checkbox.Group
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三</Checkbox>
</Checkbox.Group>
```

#### 按钮样式

普通

```jsx
<Checkbox.Group
  type="button"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三</Checkbox>
</Checkbox.Group>
```

指定默认值

```jsx
<Checkbox.Group
  type="button"
  defaultValue={['0', '1']}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三</Checkbox>
</Checkbox.Group>
```

禁用指定项

```jsx
<Checkbox.Group
  type="button"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2" disabled>选项三</Checkbox>
</Checkbox.Group>
```

圆角

```jsx
<Checkbox.Group
  type="button"
  shape="radius"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三</Checkbox>
</Checkbox.Group>
```

椭圆角

```jsx
<Checkbox.Group
  type="button"
  shape="round"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三</Checkbox>
</Checkbox.Group>
```

#### 列表样式

普通

```jsx
<Checkbox.Group
  type="cell"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2" disabled>选项三（禁止选择）</Checkbox>
</Checkbox.Group>
```

禁用状态

```jsx
<Checkbox.Group
  disabled
  type="cell"
  value={this.state.checkbox}
  onChange={value => console.log(`checkbox to ${value}`)}>
  <Checkbox value="0">选项一</Checkbox>
  <Checkbox value="1">选项二</Checkbox>
  <Checkbox value="2">选项三（禁止选择）</Checkbox>
</Checkbox.Group>
```


### API

#### Checkbox

| 属性 | 类型 | 默认值 | 可选值／参数 | 说明 |
| :--- | :--- | :--- | :--- | :--- |
| prefixCls | string | za-checkbox | | 类名前缀 |
| className | string | 无 | | 追加类名 |
| theme | string | primary | 'default', 'primary', 'info', 'success', 'warning', 'error' | 主题 |
| type | string | 无 | 'button', 'cell' | 显示类型 |
| value | string, number | 无 | | 值 |
| defaultChecked | bool | false | | 初始是否选中 |
| checked | bool | false | | 当前是否选中 |
| disabled | bool | false | | 是否禁用 |
| onChange | func | noop | \(checked : bool\) | 值变化时触发的回调函数 |

#### Checkbox.Group

| 属性 | 类型 | 默认值 | 可选值／参数 | 说明 |
| :--- | :--- | :--- | :--- | :--- |
| prefixCls | string | za-checkbox | | 类名前缀 |
| className | string | 无 | | 追加类名 |
| theme | string | primary | 'default', 'primary', 'info', 'success', 'warning', 'error' | 主题 |
| type | string | 无 | 'button', 'cell' | 显示类型 |
| shape | string | 无 | 'radius', 'round' | 形状 |
| block | bool | false | | 是否为块级元素 |
| disabled | bool | false | | 是否禁用 |
| compact | bool | false | | 是否启用紧凑模式 |
| onChange | func | noop | \(value: string, number\) | 值变化时触发的回调函数 |



