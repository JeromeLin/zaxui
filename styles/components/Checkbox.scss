$prefixCls: za-checkbox;

.#{$prefixCls} {
  display: inline-block;
  position: relative;
  margin-left: r(10);

  &:first-child {
    margin-left: 0;
  }

  &-wrapper {
    display: flex;
    align-items: center;
  }

  &-input {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
  }

  &-text {
    margin-left: r(5);
  }

  &-inner {
    position: relative;
    overflow: hidden;
    display: inline-block;
    width: r(14);
    height: r(14);
    background-color: #fff;
    transition: border-color .1s cubic-bezier(.71, -0.46, .29, 1.46), background-color .1s cubic-bezier(.71, -0.46, .29, 1.46);
    border: 1px solid $theme-primary;
    border-radius: $radius-md;

    &:before {
      content: ' ';
      position: absolute;
      left: r(4);
      top: r(1);
      width: r(4);
      height: r(8);
      border: 1px solid $theme-primary;
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg) scale(0);
      transition: all .1s cubic-bezier(.71, -0.46, .88, .6);
    }
  }

  &-group {
    &.block {
      display: flex;

      /* 通栏模式字体增大 */
      .#{$prefixCls} {
        font-size: r(15);
      }
    }

    &.shape-radius {
      .#{$prefixCls} {
        border-radius: $radius-md;
      }
    }

    &.shape-round {
      .#{$prefixCls} {
        border-radius: r(15);
      }
    }

    &.is-compact {
      .#{$prefixCls} {
        margin-left: -1px;
      }

      &.shape-radius {
        .#{$prefixCls} {
          border-radius: 0;

          &:first-child {
            border-radius: $radius-md 0 0 $radius-md;
          }

          &:last-child {
            border-radius: 0 $radius-md $radius-md 0;
          }
        }
      }

      &.shape-round {
        .#{$prefixCls} {
          border-radius: 0;

          &:first-child {
            border-radius: r(15) 0 0 r(15);
          }

          &:last-child {
            border-radius: 0 r(15) r(15) 0;
          }
        }
      }
    }
  }

  /* 选中状态 */
  &.checked {
    .#{$prefixCls}-inner {
      &:before {
        transform: rotate(45deg) scale(1);
        transition: all .2s cubic-bezier(.12, .4, .29, 1.46) .1s;
      }
    }
  }

  /* 禁用状态 */
  &.disabled {
    .#{$prefixCls}-inner {
      border-color: #d9d9d9;
      background-color: #f3f3f3;
      cursor: not-allowed;

      &:before {
        border-color: #ccc;
      }
    }
  }
}
