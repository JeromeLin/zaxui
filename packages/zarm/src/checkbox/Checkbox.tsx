import React, {
  ChangeEvent,
  InputHTMLAttributes,
  HTMLAttributes,
  ButtonHTMLAttributes,
} from 'react';
import classnames from 'classnames';
import { BaseCheckboxProps } from './interface';
import CheckboxGroup from './CheckboxGroup';
import Cell from '../cell';

const getChecked = (props: CheckboxProps, defaultChecked: boolean) => {
  return props.checked ?? props.defaultChecked ?? defaultChecked;
};

type CheckboxSpanProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type CheckboxCellProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type CheckboxButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;

export type CheckboxProps = Partial<CheckboxSpanProps & CheckboxCellProps & CheckboxButtonProps> &
  BaseCheckboxProps & {
    prefixCls?: string;
  };
interface CompoundedComponent
  extends React.ForwardRefExoticComponent<CheckboxProps & React.RefAttributes<HTMLElement>> {
  Group: typeof CheckboxGroup;
}

const Checkbox = React.forwardRef<unknown, CheckboxProps>((props, ref) => {
  const {
    prefixCls,
    className,
    type,
    shape,
    value,
    checked,
    defaultChecked,
    disabled,
    id,
    indeterminate,
    children,
    onChange,
    ...rest
  } = props;

  const [currentCheckd, setCurrentCheckd] = React.useState<boolean>(
    getChecked({ checked, defaultChecked }, false),
  );
  const checkboxRef = (ref as any) || React.createRef<HTMLElement>();

  const cls = classnames(prefixCls, className, {
    [`${prefixCls}--checked`]: currentCheckd,
    [`${prefixCls}--disabled`]: disabled,
    [`${prefixCls}--indeterminate`]: indeterminate,
    [`${prefixCls}--untext`]: !children,
  });

  const onValueChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (disabled) {
      return;
    }
    const newChecked = !currentCheckd;
    if (typeof checked === 'undefined') {
      setCurrentCheckd(newChecked);
    }
    typeof onChange === 'function' && onChange(e);
  };

  React.useEffect(() => {
    setCurrentCheckd(getChecked({ checked, defaultChecked }, false));
  }, [checked, defaultChecked]);

  const inputRender = (
    <input
      id={id}
      type="checkbox"
      className={`${prefixCls}__input`}
      value={value}
      disabled={disabled}
      checked={currentCheckd}
      onChange={onValueChange}
    />
  );

  const checkboxRender = (
    <span className={cls} ref={checkboxRef} {...(rest as CheckboxSpanProps)}>
      <span className={`${prefixCls}__widget`}>
        <span className={`${prefixCls}__inner`} />
      </span>
      {children && <span className={`${prefixCls}__text`}>{children}</span>}
      {inputRender}
    </span>
  );

  if (type === 'cell') {
    return (
      <Cell
        disabled={disabled}
        className={className}
        onClick={() => {}}
        {...(rest as CheckboxCellProps)}
      >
        {checkboxRender}
      </Cell>
    );
  }

  if (type === 'button') {
    return (
      <button
        ref={checkboxRef}
        type="button"
        disabled={disabled}
        className={cls}
        {...(rest as CheckboxButtonProps)}
      >
        {children}
        {inputRender}
      </button>
    );
  }

  return checkboxRender;
}) as CompoundedComponent;

Checkbox.displayName = 'Checkbox';

Checkbox.defaultProps = {
  prefixCls: 'za-checkbox',
  disabled: false,
  indeterminate: false,
};

export default Checkbox;
