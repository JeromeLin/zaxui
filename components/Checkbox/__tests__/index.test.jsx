import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Checkbox from '../index';

describe('Checkbox', () => {
  it('renders correctly', () => {
    const wrapper = render(<Checkbox checked>foo</Checkbox>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('checked change false', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Checkbox checked onChange={onChange}>foo</Checkbox>);
    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(onChange).toHaveBeenCalledWith(false);
  });
});
