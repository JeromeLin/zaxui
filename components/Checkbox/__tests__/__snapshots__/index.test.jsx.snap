// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<div
  class="za-checkbox checked"
>
  <div
    class="za-checkbox-wrapper"
  >
    <span
      class="za-checkbox-inner"
    />
    <span
      class="za-checkbox-text"
    >
      foo
    </span>
    <input
      checked=""
      class="za-checkbox-input"
      type="checkbox"
    />
  </div>
</div>
`;
